File: plugins/mzizzi/skills/codex-adversarial-review/scripts/tests/test_run_review.py
#!/usr/bin/env python3
"""Regression checks for run_review.py's failure contract: the underlying script's exit
status must propagate, stdout must be released only on success, stderr only on failure.

Usage: python test_run_review.py     (exits non-zero if any case fails)
"""

import json
import os
import subprocess
import sys
import tempfile
from pathlib import Path

script = Path(__file__).resolve().parents[1] / "run_review.py"

with tempfile.TemporaryDirectory() as tmp:
    work = Path(tmp)

    # A fake home so plugin resolution succeeds and the run reaches the part under test.
    fake_home = work / "home"
    # A real .mjs run by the real node, not a stub `node` earlier on PATH: Windows
    # CreateProcess appends only .exe and never consults PATHEXT, so a stub is skipped.
    companion = fake_home / "plug" / "scripts" / "codex-companion.mjs"
    companion.parent.mkdir(parents=True)
    (fake_home / ".claude" / "plugins").mkdir(parents=True)
    (fake_home / ".claude" / "plugins" / "installed_plugins.json").write_text(
        json.dumps({"plugins": {"codex@openai-codex": [{"installPath": str(companion.parents[1])}]}}),
        encoding="utf-8",
    )

    env = dict(os.environ)
    # Path.home() reads USERPROFILE on Windows and HOME everywhere else, so both point at
    # the fake home to keep this test platform-agnostic.
    env["HOME"] = env["USERPROFILE"] = str(fake_home)

    fails = 0
    for code in (0, 1, 2, 7):
        # Writes to stdout *before* failing, mimicking the companion rendering a partial
        # report and only then settling on a non-zero status.
        companion.write_text(
            f'process.stdout.write("REPORT BODY (stub exit {code})\\n");\n'
            f'process.stderr.write("stub diagnostic for exit {code}\\n");\n'
            f"process.exit({code});\n",
            encoding="utf-8",
        )

        completed = subprocess.run(
            [sys.executable, str(script), "--scope auto test"],
            capture_output=True,
            text=True,
            env=env,
        )
        out = completed.stdout.strip()
        err = completed.stderr.strip()

        want_out = f"REPORT BODY (stub exit {code})" if code == 0 else ""
        want_err = "" if code == 0 else f"stub diagnostic for exit {code}"

        ok = completed.returncode == code and out == want_out and err == want_err
        fails += not ok

        print(
            f"underlying exit {code:<2} -> wrapper {completed.returncode:<2}  "
            f"stdout={'nonempty' if out else 'empty':<9} "
            f"stderr={'nonempty' if err else 'empty':<9} "
            f"[{'yes' if ok else 'no'}]"
        )

print()
if fails:
    sys.exit(f"FAIL: {fails} case(s)")
print("PASS")

File: plugins/mzizzi/skills/codex-adversarial-review/scripts/tests/test_run_review.mjs
#!/usr/bin/env node
// Regression checks for run_review.mjs: --files must scope the review input to those paths, an
// unusable --files must stop rather than fall back to the whole tree, and an empty report must
// retry once before failing.
//
// Usage: node test_run_review.mjs     (exits non-zero if any case fails)

import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";

const script = path.join(path.dirname(fileURLToPath(import.meta.url)), "..", "run_review.mjs");
const tmp = fs.mkdtempSync(path.join(os.tmpdir(), "run-review-test-"));
const home = path.join(tmp, "home");
const plugin = path.join(home, "plug");
const repo = path.join(tmp, "repo");
const promptLog = path.join(tmp, "prompts.txt");

const write = (file, body) => {
  fs.mkdirSync(path.dirname(file), { recursive: true });
  fs.writeFileSync(file, body, "utf8");
};

// Stubs standing in for the codex plugin's lib modules. runAppServerTurn records the prompt it
// was handed, which is what the scoping assertions read.
write(
  path.join(plugin, "scripts", "lib", "git.mjs"),
  `export function ensureGitRepository() {}
export function getRepoRoot() { return process.env.TEST_REPO; }
export function getCurrentBranch() { return "main"; }
export function resolveReviewTarget() { return { mode: "working-tree", label: "working tree diff" }; }
export function collectReviewContext() {
  return { repoRoot: process.env.TEST_REPO, branch: "main", target: { label: "working tree diff" },
           collectionGuidance: "guidance", content: "WHOLE_TREE_MARKER" };
}\n`
);
write(
  path.join(plugin, "scripts", "lib", "codex.mjs"),
  `import fs from "node:fs";
export function readOutputSchema() { return {}; }
export async function runAppServerTurn(cwd, options) {
  fs.appendFileSync(process.env.TEST_PROMPT_LOG, options.prompt + "\\n=====\\n");
  return { finalMessage: process.env.TEST_EMPTY ? "" : '{"verdict":"approve"}', status: "ok" };
}
export function parseStructuredOutput(raw) {
  return raw ? { parsed: JSON.parse(raw), rawOutput: raw } : { parsed: null, rawOutput: "" };
}\n`
);
write(
  path.join(plugin, "scripts", "lib", "prompts.mjs"),
  `import fs from "node:fs";
import path from "node:path";
export function loadPromptTemplate(root, name) {
  return fs.readFileSync(path.join(root, "prompts", name + ".md"), "utf8");
}
export function interpolateTemplate(template, vars) {
  return template.replace(/\\{\\{([A-Z_]+)\\}\\}/g, (_, k) => (k in vars ? vars[k] : ""));
}\n`
);
write(
  path.join(plugin, "scripts", "lib", "render.mjs"),
  `export function renderReviewResult(parsed, meta) { return "RENDERED " + meta.targetLabel + "\\n"; }\n`
);
write(path.join(plugin, "prompts", "adversarial-review.md"), "INPUT={{REVIEW_INPUT}}\n");
write(path.join(plugin, "schemas", "review-output.schema.json"), "{}\n");
write(
  path.join(home, ".claude", "plugins", "installed_plugins.json"),
  JSON.stringify({ plugins: { "codex@openai-codex": [{ installPath: plugin }] } })
);

write(path.join(repo, "target.md"), "TARGET_MARKER\n");
write(path.join(repo, "other.md"), "OTHER_MARKER\n");
spawnSync("git", ["init", "-q"], { cwd: repo });

function run(args, { empty = false, noPlugin = false } = {}) {
  fs.writeFileSync(promptLog, "");
  const env = {
    ...process.env,
    HOME: noPlugin ? path.join(tmp, "nohome") : home,
    USERPROFILE: noPlugin ? path.join(tmp, "nohome") : home,
    TEST_REPO: repo,
    TEST_PROMPT_LOG: promptLog
  };
  if (empty) env.TEST_EMPTY = "1";
  else delete env.TEST_EMPTY;

  const result = spawnSync(process.execPath, [script, ...args], { cwd: repo, encoding: "utf8", env });
  return { ...result, prompts: fs.readFileSync(promptLog, "utf8") };
}

const cases = [
  [
    "--files scopes the input to that file",
    () => {
      const r = run(["--files", "target.md", "focus"]);
      return (
        r.status === 0 &&
        r.stdout.includes("RENDERED files: target.md") &&
        r.prompts.includes("TARGET_MARKER") &&
        !r.prompts.includes("OTHER_MARKER") &&
        !r.prompts.includes("WHOLE_TREE_MARKER")
      );
    }
  ],
  [
    "no --files falls back to the plugin's own collector",
    () => run(["focus"]).prompts.includes("WHOLE_TREE_MARKER")
  ],
  [
    "--files resolving to nothing stops instead of widening",
    () => {
      const r = run(["--files", " , ", "focus"]);
      return r.status === 2 && !r.prompts.includes("WHOLE_TREE_MARKER");
    }
  ],
  ["--files naming a missing path exits 2", () => run(["--files", "nope.md"]).status === 2],
  [
    "an empty report retries once, then fails",
    () => {
      const r = run(["--files", "target.md"], { empty: true });
      return (
        r.status === 1 &&
        r.stderr.includes("CODEX_EMPTY_REPORT") &&
        r.prompts.split("=====").length - 1 === 2
      );
    }
  ],
  [
    "a missing plugin exits 2",
    () => {
      const r = run(["--files", "target.md"], { noPlugin: true });
      return r.status === 2 && r.stderr.includes("CODEX_NOT_INSTALLED");
    }
  ]
];

let failed = 0;
for (const [name, check] of cases) {
  const ok = check();
  failed += !ok;
  console.log(`${ok ? "pass" : "FAIL"}  ${name}`);
}

fs.rmSync(tmp, { recursive: true, force: true });
console.log();
if (failed) {
  console.error(`FAIL: ${failed} case(s)`);
  process.exit(1);
}
console.log("PASS");
